lib/route/ROUTING.md

```diff
@@ -0,0 +1,446 @@
+# Obelisk Routing
+
+The `obelisk-routes` package is designed to help with managing the paths and parameters for
+routing in your application.
+
+This is distinct from packages like `servant` that are designed for specifying endpoints for a
+REST API. When building routes with `obelisk-routes` you're not going to be handling headers or
+HTTP methods.
+
+One of the goals of `obelisk-routes` is to allow for complex routing structures to be partitioned
+into manageable chunks that are then combined in a straight-forward manner. Allowing for re-use
+of various components and to help the type-system provide as many guarantees as possible. We'll
+see this come into play when we start adding more complex routes.
+
+## Building blocks
+
+There are a few pieces required to build routes in Obelisk and there is plenty to dig into if you
+are curious. It may seem daunting and there is certainly a lot of power on offer. However for our
+purposes we can survive with a surface level understanding.
+
+Below we will cover a few of the types that we need to interact with to add routes to our
+application. Again you don't have to have a deep understanding of these types, a high level view
+will suffice for now.
+
+#### `SegmentResult`
+
+This is the primary building block of individual routes, it has a list-like structure with a
+recursive component, and a terminating case.
+
+The recursive case is the `PathSegment Text (Encoder check parse a PageName)` constructor that
+encodes a literal segment in the path, followed by one or more `Encoder`s.
+
+The terminating case is the `PathEnd (Encoder check parse a (Map Text (Maybe Text)))` constructor.
+It is used to indicate that there is only this single segment in that route, or to terminate a longer route.
+
+#### `Encoder`
+
+An `Encoder` may describe how an individual segment of a route is to be encoded. Such as how to
+read the number from a route like `/user/33`. Or how several routes are encoded. So you may have
+an `Encoder` that for all of the routes pertaining to 'blog posts', such as:
+
+- /blog/:postId:/edit
+- /blog/:postId:/metrics
+- etc...
+
+In particular note that `Encoder`s may be composed, thereby creating more elaborate routing
+structures from simpler pieces. It also means that scary looking routing hierarchies can be
+partitioned and made easier to work with.
+
+#### `BackendRoute` & `FrontendRoute`
+
+An Obelisk application begins with two distinct sets of routes for backend and frontend. These
+are defined as GADTs called `BackendRoute` and `FrontendRoute`, respectively. These are provided
+for you to start you off.
+
+In a fresh Obelisk application these live in the `Common.Route` module.
+
+`BackendRoute` is for serving static things or deferring to other backend endpoints. Such as:
+- "static/files/img/:imgName:"
+- "api/..." where the "..." portion is managed by a REST endpoint package of your choice.
+
+`FrontendRoute` is for managing the frontend routing of your application. Such as:
```

Comment:
Even if it's slightly redundant, it might be helpful to add a note around here something like "These are all the routes for which the backend will serve up the application frontend."

Reply:
I wanted to expand that sentence a bit as I wasn't quite able to parse it without a bit more description. :)

> These backend routes can run their own code or hand off to a different request handling package
> to a finer grained set of endpoints. But should none of them match the frontend of the
> application is served and the frontend router will continue parsing the route.